src/puppetlabs/puppetdb/scf/migrate.clj

```diff
@@ -1124,17 +974,329 @@
    "ALTER TABLE fact_paths DROP COLUMN value_type_id"
    "ALTER TABLE fact_values DROP COLUMN path_id"))
 
-(defn fill-in-null-producer-timestamp
-  "The producer_timestamp column can sometimes contain null values,
-  especially when updating from older versions. New use cases require
-  a value in that position, so we'll use the value from the
-  'timestamp' column if it's not there. "
+(defn version-2yz-to-300-migration
+  ;; This migration includes:
+  ;;   Insertion of the factsets hash column
+  ;;   Using a report_id (surrogate key) instead of the reports.hash (natural key)
+  ;;   Insert a noop column into reports
+  ;;   Drop latest_report table, merge with certnames
+  ;;   Change name to certname where inconsistent
+  ;;   Insert reports metrics and logs
+  ;;   Ensuring producer_timestamp is NOT NULL
+  ;;   Changing the types of hashes and uuids in postgres to bytea and uuid respectively
```

Comment:
I don't want to beat a dead horse on this -- the collapsed migration seems alright as far as I can see. In cases like these, absent a real performance benefit, my tendency would be to keep things discrete and duplicate code as necessary to do so. The fact is that code is heavily duplicated across all of our migrations, since generally they happen across releases and refactoring existing ones is usually not an option.